extensions/sql-database-projects/src/controllers/databaseProjectTreeViewProvider.ts

```diff
@@ -0,0 +1,132 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the Source EULA. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as vscode from 'vscode';
+import { promises as fs } from 'fs';
+import * as path from 'path';
+import * as nls from 'vscode-nls';
+
+import { SqlDatabaseProjectItem } from './databaseProjectTreeItem';
+import { Deferred } from '../common/promise';
+
+
+const localize = nls.loadMessageBundle();
+
+export class SqlDatabaseProjectTreeViewProvider implements vscode.TreeDataProvider<SqlDatabaseProjectItem> {
+	private _onDidChangeTreeData: vscode.EventEmitter<SqlDatabaseProjectItem | undefined> = new vscode.EventEmitter<SqlDatabaseProjectItem | undefined>();
+	readonly onDidChangeTreeData: vscode.Event<SqlDatabaseProjectItem | undefined> = this._onDidChangeTreeData.event;
+
+	private _initializeDeferred: Deferred<void> = new Deferred<void>();
+
+	private roots: SqlDatabaseProjectItem[] = [];
+
+	constructor() {
+		this._initializeDeferred = new Deferred<void>();
+		this.initialize();
+	}
+
+	async initialize() {
+		this.roots = [new SqlDatabaseProjectItem(localize('noProjectOpenMessage', "No open database project"), false)];
+
+
+		this._initializeDeferred.resolve();
+	}
+
+	getTreeItem(element: SqlDatabaseProjectItem): vscode.TreeItem {
+		return {
+			label: element.label,
+			collapsibleState: element.parent === undefined
+				? vscode.TreeItemCollapsibleState.Expanded
+				: element.isFolder
+					? vscode.TreeItemCollapsibleState.Collapsed
+					: vscode.TreeItemCollapsibleState.None
+		};
+	}
+
+	getChildren(element?: SqlDatabaseProjectItem): SqlDatabaseProjectItem[] {
+		if (element === undefined) {
+			return this.roots;
+		}
+
+		return element.children;
+	}
+
+	public get initialized(): Promise<void> {
+		return this._initializeDeferred.promise;
+	}
+
+	async openProject(projectFiles: vscode.Uri[]) {
+		if (projectFiles.length > 1) { // TODO: how to handle opening a folder with multiple .sqlproj files?
+			vscode.window.showErrorMessage('Multiple .sqlproj files selected; please select only one.');
+		}
+
+		if (projectFiles.length === 0) {
+			vscode.window.showErrorMessage('No .sqlproj file selected; please select one.');
+		}
+
+		let directoryPath = path.dirname(projectFiles[0].fsPath);
```

Comment:
This is going to throw an exception if projectFiles.length === 0. You check for that above but don't return so it'll continue on to this - the above checks should probably  be returning.